### PR TITLE
[PropertyAccess] Backport fixes from 2.7

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch? | 2.3 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

While merging 2.3 into 2.7, I noticed that many fixes were merged into 2.7 that should have been merged into 2.3. This PR backports them, with a few other fixes applied when merging 2.3 into 2.7.

### CODE REVIEW

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 1]
```diff
@@ -187,19 +226,21 @@ public function setValue(&$objectOrArray, $propertyPath, $value)
...
+    private static function throwInvalidArgumentException($message, $trace, $i)
     {
         if (isset($trace[$i]['file']) && __FILE__ === $trace[$i]['file']) {
             $pos = strpos($message, $delim = 'must be of the type ') ?: strpos($message, $delim = 'must be an instance of ');
             $pos += strlen($delim);
+            $type = $trace[$i]['args'][0];
+            $type = is_object($type) ? get_class($type) : gettype($type);
 
-            throw new UnexpectedTypeException($trace[$i]['args'][0], substr($message, $pos, strpos($message, ',', $pos) - $pos));
+            throw new InvalidArgumentException(sprintf('Expected argument of type "%s", "%s" given', substr($message, $pos, strpos($message, ',', $pos) - $pos), $type));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: Using UnexpectedTypeException in not compatible with 2.7, where this exception type has been specialized. Using InvalidArgumentException backported from 2.7 instead.

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 2]
```diff
@@ -137,7 +156,7 @@ public function setValue(&$objectOrArray, $propertyPath, $value)
         $overwrite = true;
 
         try {
-            if (PHP_VERSION_ID < 70000) {
+            if (PHP_VERSION_ID < 70000 && false === self::$previousErrorHandler) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: this is a bug fix when a property accessor is callinged an other property accessor

**Comment by dunglas** [comment]
:+1:

**Comment by stof** [comment]
:+1:

**Pull request merged**